aws/resource_aws_lambda_layer_test.go

```diff
@@ -0,0 +1,215 @@
+	}
+
+	if len(resp.Layers) == 0 {
+		log.Print("[DEBUG] No aws lambda layers to sweep")
+		return nil
+	}
+
+	for _, l := range resp.Layers {
+		var testOptGroup bool
+		for _, testName := range []string{"tf_test", "tf_acc_"} {
+			if strings.HasPrefix(*l.LayerName, testName) {
+				testOptGroup = true
+			}
+		}
+
+		if !testOptGroup {
+			continue
+		}
+
+		versionResp, err := lambdaconn.ListLayerVersions(&lambda.ListLayerVersionsInput{
+			LayerName: l.LayerName,
+		})
+		if err != nil {
+			return fmt.Errorf("Error retrieving versions for lambda layer: %s", err)
+		}
+
+		for _, v := range versionResp.LayerVersions {
+			_, err := lambdaconn.DeleteLayerVersion(&lambda.DeleteLayerVersionInput{
+				LayerName:     l.LayerName,
+				VersionNumber: v.Version,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func TestAccAWSLambdaLayer_basic(t *testing.T) {
+	resourceName := "aws_lambda_layer.lambda_layer_test"
+	layerName := fmt.Sprintf("tf_acc_lambda_layer_basic_%s", acctest.RandString(8))
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckLambdaLayerDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSLambdaLayerBasic(layerName),
+				Check:  testAccCheckAwsLambdaLayerExists(resourceName, layerName),
+			},
+
+			{
+				ResourceName:            resourceName,
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"filename"},
+			},
+		},
+	})
+}
+
+func TestAccAWSLambdaLayer_s3(t *testing.T) {
+	resourceName := "aws_lambda_layer.lambda_layer_test"
+	rString := acctest.RandString(8)
+	layerName := fmt.Sprintf("tf_acc_lambda_layer_s3_%s", rString)
+	bucketName := fmt.Sprintf("tf-acc-bucket-lambda-layer-s3-%s", rString)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckLambdaLayerDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSLambdaLayerS3(bucketName, layerName),
+				Check:  testAccCheckAwsLambdaLayerExists(resourceName, layerName),
+			},
+
+			{
+				ResourceName:            resourceName,
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"s3_bucket", "s3_key"},
+			},
+		},
+	})
+}
+
+func testAccCheckLambdaLayerDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).lambdaconn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_lambda_layer" {
+			continue
+		}
+
+		version, err := strconv.Atoi(rs.Primary.Attributes["version"])
+		if err != nil {
+			return err
+		}
+
+		_, err = conn.GetLayerVersion(&lambda.GetLayerVersionInput{
+			LayerName:     aws.String(rs.Primary.ID),
+			VersionNumber: aws.Int64(int64(version)),
+		})
+		if err == nil {
```

Comment:
We generally prefer to invert this conditional because its important to always return errors if we receive them, e.g.

```go
if isAWSErr(err, lambda.ErrCodeResourceNotFoundException, "") {
	continue
}

if err != nil {
	return err
}

return fmt.Errorf("Lambda Layer Version (%s) still exists", rs.Primary.ID)
```